Overriding executable `rpath` with patchelf

mmaslankaprv opened the pull request:
Overriding Redpanda executable makes it easier to handle shared library loading.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
bazel: add patchelf module

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
bazel: added ability to override rpath in packaging

By overriding an `executable` rpath we no longer need to care about
setting correct `LB_LIBRARY_PATH`

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
bazel: override rpath in bazel packages

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
bazel/packaging: added overriding interpreter path

Signed-off-by: Michał Maślanka <michal@redpanda.com>

rockwotj commented on bazel/packaging/packaging.bzl:
```diff
@@ -18,17 +18,54 @@ def _is_versioned_so(file):
+        inputs = [original_binary],
+        outputs = [patched_binary],
+        executable = ctx.executable._patchelf,
+        arguments = ["--force-rpath", "--set-rpath", path_override, original_binary.path, "--output", patched_binary.path],
```
Why are we setting `--force-rpath`? https://man.archlinux.org/man/patchelf.1.en#force-rpath

mmaslankaprv replied:
Actually we do not need it, fixed

rockwotj commented on bazel/packaging/bin_wrapper/BUILD:
```diff
@@ -17,7 +17,6 @@ go_binary(
     visibility = ["//bazel/packaging:__pkg__"],
     x_defs = {
         "binaryPath": "/opt/redpanda/libexec/redpanda",
-        "ldLibraryPath": "/opt/redpanda/lib",
```
Do we even need these anymore with this approach? Should we remove the wrappers and just put the real binaries in bin/ instead of libexec/?

rockwotj reviewed (dismissed):
Amazing thank you 🙏 

Just a couple of questions for my understanding

mmaslankaprv committed:
bazel/packaging: removed bin_wrapper scripts

Removed binary wrapper scripts as they are not longer required with
patched executables.

Signed-off-by: Michał Maślanka <michal@redpanda.com>
